### PR TITLE
8345801: C2: Clean up include statements to speed up compilation when touching type.hpp

### PR DESCRIPTION
**Pull request by chhagedorn**
I've noticed that after touching `type.hpp` in Valhalla, it requires more than 7 minutes to build hotspot again on my machine. I would have suspected that we mostly recompile C2/opto source files. But that is far from the truth: A lot of unrelated source files must be recompiled, including, for example, C1, JFR, or runtime files.

In mainline, the impact is not that severe. But it still requires around 1 minute to build hotspot again on my machine after touching `type.hpp`. I've had a look at the include chains and removed quite a lot of unused includes. For the active includes, the most impact had including `output.hpp` inside `c2compiler.hpp`. This set up the following dependency chain:
```
... more C1 files
  #include "c1/c1_Compilation.hpp"
    #include "compiler/compilerDefinitions.inline.hpp"
      #include "opto/c2compiler.hpp"
        #include "opto/output.hpp"    <------------ Problematic include
          #include "opto/c2_CodeStubs.hpp"
            #include "opto/compile.hpp"
              ... more opto files and eventually type.hpp
```
This means that a lot of C1 files also need to be re-compiled as well as some more source file that include `compiler/compilerDefinitions.inline.hpp`. I cut this dependency chain by removing the include of `opto/output.hpp` in `opto/c2compiler.hpp` and moved the constant `initial_const_capacity` from `output.hpp` to `c2Compiler.hpp` which seemed to be the only reason why we have the include in place. After this change, I was required to add some missing includes that were included transitively before.

The final mainline patch could also be applied to the current Valhalla repository (with some small tweaks). The results were quite promising. I could bring the compilation time on my machine significantly down in mainline and especially in Valhalla after touching `type.hpp`:

- Mainline: ~1min -> ~40s (1.5 times faster)
- Valhalla: ~7min -> ~40s (10.5 times faster)

I've only focused on `type.hpp` here but I guess other includes in the JIT compiler area or other parts of hotspot could also be revisited to possibly speed up the compilation after touching some header files.

Testing:
- Oracle CI
- GHA

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8345801](https://bugs.openjdk.org/browse/JDK-8345801): C2: Clean up include statements to speed up compilation when touching type.hpp (**Enhancement** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22658/head:pull/22658` \
`$ git checkout pull/22658`

Update a local copy of the PR: \
`$ git checkout pull/22658` \
`$ git pull https://git.openjdk.org/jdk.git pull/22658/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22658`

View PR using the GUI difftool: \
`$ git pr show -t 22658`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22658.diff">https://git.openjdk.org/jdk/pull/22658.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22658#issuecomment-2531019539)
</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Good.

File: src/hotspot/share/opto/node.hpp [hunk 1]
```diff
@@ -2000,6 +2000,10 @@ Compile::locate_node_notes(GrowableArray<Node_Notes*>* arr,
   return arr->at(block_idx) + (idx & (_node_notes_block_size-1));
 }
 
+inline Node_Notes* Compile::node_notes_at(int idx) {
```
Comment (dean-long) [inline src/hotspot/share/opto/node.hpp:2003]: It seems weird to have these inlined Compile methods in node.hpp.  Could we clean this up separately, maybe by moving them to compile.inline.hpp?

**Review by dean-long — APPROVED** [review]
Looks good, as long as you tested w/ and w/o precompiled headers.

**Review by TheShermanTanker — APPROVED** [review]
It takes just 7 minutes to compile and link HotSpot for you? If only I had that kind of luxury...

**Comment by chhagedorn** [comment]
Thanks @TheShermanTanker for your review! The measured numbers are only when starting a hotspot build again after touching `type.hpp`. It takes slightly longer for a full `make hotspot`.

**Pull request closed without merging**